### PR TITLE
INGK-631 Add mypy

### PR DESCRIPTION
**Pull request by pzinemanas**
### Description

This PR add type hints and mypy checking.

Fixes # 631

### Type of change

- [x] Add type hints
- [x] Make some changes to simplify the type checking
- [x] Add mypy check

### Tests
- [x] Change tests to adapt them to the changes.
- [x] Run tests.

### Documentation

Please update the documentation.

- [x] Update docstrings of every function, method or class that change.
- [x] Build documentation locally to verify changes.
- [ ] Add the changes at the `[Unreleased]` section of the [CHANGELOG](CHANGELOG.md).

### Code formatting

- [x] Use black package to format the code: `black -l 100 ingenialink tests`. It is recommended to configure the code editor to automatically format the code using black with a max length line of 100.

### CODE REVIEW

File: ingenialink/canopen/dictionary.py [hunk 1]
```diff
@@ -14,35 +18,42 @@ class CanopenDictionary(Dictionary):
 
     """
 
-    class AttrRegCanDict(Dictionary.AttrRegDict):
-        IDX = "idx"
-        SUBIDX = "subidx"
-
-    def __init__(self, dictionary_path):
+    def __init__(self, dictionary_path: str) -> None:
```
Comment (MartinALL91) [inline ingenialink/canopen/dictionary.py:21]: Remove type from docstring

Comment (MartinALL91) [reply ingenialink/canopen/dictionary.py:21]: This comment applies to all functions that you changed.

File: ingenialink/canopen/register.py [hunk 2]
```diff
@@ -33,23 +35,25 @@ class CanopenRegister(Register):
...
-        enums=None,
-        cat_id=None,
-        scat_id=None,
-        internal_use=0,
-        address_type=None,
+        idx: int,
+        subidx: int,
+        dtype: REG_DTYPE,
+        access: REG_ACCESS,
+        identifier: Optional[str],
```
Comment (MartinALL91) [inline ingenialink/canopen/register.py:42]: Default values are missing

File: ingenialink/canopen/servo.py [hunk 3]
```diff
@@ -51,18 +54,24 @@ class CanopenServo(Servo):
         subnode=0,
     )
 
-    def __init__(self, target, node, dictionary_path=None, servo_status_listener=False):
+    def __init__(
+        self,
+        target: int,
+        node: canopen.RemoteNode,
+        dictionary_path: str,
```
Comment (MartinALL91) [inline ingenialink/canopen/servo.py:61]: Same as https://github.com/ingeniamc/ingenialink-python/pull/319/files#r1364142793

File: ingenialink/ethernet/dictionary.py [hunk 4]
```diff
@@ -14,32 +17,52 @@ class EthernetDictionary(Dictionary):
...
         super().__init__(dictionary_path)
 
-    def _read_xdf_register(self, register):
+    def _read_xdf_register(self, register: ET.Element) -> Optional[EthernetRegister]:
         current_read_register = super()._read_xdf_register(register)
         if current_read_register is None:
             return None
         try:
-            current_read_register[self.AttrRegEthDict.ADDR] = int(register.attrib["address"], 16)
+            address = int(register.attrib["address"], 16)
```
Comment (GullinBustin) [inline ingenialink/ethernet/dictionary.py:29]: Does not use "address" as a literal

File: ingenialink/dictionary.py [hunk 5]
```diff
@@ -63,16 +71,18 @@ class DictionaryErrors:
...
 
         self.load_errors()
 
-    def load_errors(self):
+    def load_errors(self) -> None:
         """Load errors from dictionary."""
         for element in self._list_xdf_errors:
             label = element.find(DICT_LABELS_LABEL)
+            if label is None:
+                continue
```
Comment (GullinBustin) [inline ingenialink/dictionary.py:85]: Log a warning

File: ingenialink/dictionary.py [hunk 6]
```diff
@@ -25,25 +27,31 @@ class DictionaryCategories:
...
         """Load category IDs from dictionary."""
         for element in self._list_xdf_categories:
             self._cat_ids.append(element.attrib["id"])
-            self._categories[element.attrib["id"]] = {"en_US": element.find(DICT_LABELS_LABEL).text}
+            cat_element = element.find(DICT_LABELS_LABEL)
+            if cat_element is None:
+                continue
+            cat_id = cat_element.text
+            if cat_id is None:
+                continue
```
Comment (GullinBustin) [inline ingenialink/dictionary.py:46]: Log a warning in both checks

File: ingenialink/dictionary.py [hunk 7]
```diff
@@ -128,21 +138,6 @@ class Dictionary(ABC):
     DICT_ENUMERATIONS = "./Enumerations"
     DICT_ENUMERATIONS_ENUMERATION = f"{DICT_ENUMERATIONS}/Enum"
 
-    class AttrRegDict:
```
Comment (GullinBustin) [inline ingenialink/dictionary.py:131]: Why did you remove this? The idea is avoid use literal strings

**Review by GullinBustin — CHANGES_REQUESTED** [review]

File: ingenialink/canopen/dictionary.py (hunk 1 above)
Comment (pzinemanas) [reply ingenialink/canopen/dictionary.py:20]: Done!

File: ingenialink/canopen/register.py (hunk 2 above)
Comment (pzinemanas) [reply ingenialink/canopen/register.py]: Which default values are missing?

File: ingenialink/dictionary.py (hunk 7 above)
Comment (pzinemanas) [reply ingenialink/dictionary.py:131]: This was used to save the keys of the dictionary used to fill the register attributes before creating a Register instance. As we are not doing so anymore, this enum is not used.

File: ingenialink/canopen/network.py [hunk 8]
```diff
@@ -307,15 +319,19 @@ def _setup_connection(self):
         else:
             logger.info("Connection already established")
 
-    def _teardown_connection(self):
+    def _teardown_connection(self) -> None:
         """Tears down the already established connection
         and deletes the network interface"""
+        if self._connection is None:
+            return
```
[the remark this answers is not in the document]
Comment (pzinemanas) [reply ingenialink/canopen/network.py:327]: Done!

File: ingenialink/canopen/network.py [hunk 9]
```diff
@@ -307,15 +319,19 @@ def _setup_connection(self):
...
         self._connection.disconnect()
         self._connection = None
         logger.info("Tear down connection.")
 
-    def _reset_connection(self):
+    def _reset_connection(self) -> None:
         """Resets the established CANopen network."""
+        if self._connection is None:
+            return
```
[the remark this answers is not in the document]
Comment (pzinemanas) [reply ingenialink/canopen/network.py]: Done!

File: ingenialink/canopen/network.py [hunk 10]
```diff
@@ -446,7 +462,9 @@ def __force_boot(
             callback_status_msg: Subscribed callback function for the status message
 
         """
-        device_type = servo.read(CIA301_DRV_ID_DEVICE_TYPE, subnode=0)
+        if self._connection is None:
+            return
```
[the remark this answers is not in the document]
Comment (pzinemanas) [reply ingenialink/canopen/network.py]: Done!

File: ingenialink/canopen/network.py [hunk 11]
```diff
@@ -798,20 +830,24 @@ def change_node_id(
             self._lss_reset_connection_nodes(target_node)
             logger.info("Node ID changed to {}".format(new_target_node))
 
-    def _lss_store_configuration(self):
+    def _lss_store_configuration(self) -> None:
         """Stores the current configuration of the LSS"""
+        if self._connection is None:
+            return
```
[the remark this answers is not in the document]
Comment (pzinemanas) [reply ingenialink/canopen/network.py]: Done!

File: ingenialink/canopen/network.py [hunk 12]
```diff
@@ -798,20 +830,24 @@ def change_node_id(
...
-    def _lss_reset_connection_nodes(self, target_node):
+    def _lss_reset_connection_nodes(self, target_node: int) -> None:
         """Resets the connection and starts node guarding for the connection nodes.
 
         Args:
             target_node (int): Node ID of the targeted device.
 
         """
+        if self._connection is None:
+            return
```
[the remark this answers is not in the document]
Comment (pzinemanas) [reply ingenialink/canopen/network.py]: Done!

File: ingenialink/canopen/network.py [hunk 13]
```diff
@@ -869,8 +905,10 @@ def start_status_listener(self, servo):
         listener.start()
         self.__listeners_net_status.append(listener)
 
-    def stop_status_listener(self, servo):
+    def stop_status_listener(self, servo: CanopenServo) -> None:  # type: ignore [override]
         """Stops the NetStatusListener from listening to the drive."""
+        if self._connection is None:
+            return
```
[the remark this answers is not in the document]
Comment (pzinemanas) [reply ingenialink/canopen/network.py]: Done!

File: ingenialink/ethernet/dictionary.py (hunk 4 above)
Comment (pzinemanas) [reply ingenialink/ethernet/dictionary.py]: Done!

File: ingenialink/dictionary.py (hunk 5 above)
Comment (pzinemanas) [reply ingenialink/dictionary.py:88]: Done!

File: ingenialink/dictionary.py (hunk 6 above)
Comment (pzinemanas) [reply ingenialink/dictionary.py:48]: Done!

File: ingenialink/canopen/register.py (hunk 2 above)
Comment (MartinALL91) [reply ingenialink/canopen/register.py]: It was fixed in this [commit ](https://github.com/ingeniamc/ingenialink-python/pull/319/commits/a89c27f89a5185876cea5e4768d07c9d7a66bfa0)

File: ingenialink/dictionary.py (hunk 7 above)
Comment (GullinBustin) [reply ingenialink/dictionary.py:131]: But now you always use literals for the keys used for "register.attrib", if some day dictionary keys change we must look for all the literals and replace them, It's better to use constants, enums, or something like that.

File: ingenialink/ethernet/dictionary.py [hunk 14]
```diff
@@ -1,7 +1,10 @@
+from typing import Optional, Dict, List
+
 from ingenialink.dictionary import Dictionary
 from ingenialink.ethernet.register import EthernetRegister
 
 import ingenialogger
+import xml.etree.ElementTree as ET
```
Comment (GullinBustin) [inline ingenialink/ethernet/dictionary.py:7]: Imports orders should be:

1. Standard library imports.
2. Related third party imports.
3. Local application/library specific imports.

https://peps.python.org/pep-0008/#imports

File: ingenialink/virtual/virtual_drive.py [hunk 15]
```diff
@@ -22,91 +24,119 @@ class VirtualMonDistBase:
...
-    DATA_REG = None
-
-    def __init__(self, drive):
+    FREQ_DIVIDER_REG = ""
+    BUFFER_SIZE_REG = ""
+    NUMBER_MAP_REGS = ""
+    MAP_REG_CFG = ""
+    BYTES_PER_BLOCK_REG = ""
+    AVAILABLE_BYTES_REG = ""
+    DATA_REG = ""
```
Comment (GullinBustin) [inline ingenialink/virtual/virtual_drive.py:38]: Maybe you can simply write `DATA_REG: str`. Same with the others

File: mypy.ini [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+ignore_missing_imports = True
+
+[mypy-can]
+ignore_missing_imports = True
+
+[mypy-ingenialogger]
+ignore_missing_imports = True
+
+[mypy-numpy]
+ignore_missing_imports = True
```
Comment (GullinBustin) [inline mypy.ini:22]: where is numpy used?

**Review by GullinBustin — CHANGES_REQUESTED** [review]

File: ingenialink/dictionary.py (hunk 7 above)
Comment (pzinemanas) [reply ingenialink/dictionary.py:131]: Yes, but this class was not used for that purpose. Would you like to use it to store the xdf keys?

File: mypy.ini (hunk 16 above)
Comment (pzinemanas) [reply mypy.ini:22]: It is not used. I'll remove it.

File: ingenialink/ethernet/dictionary.py (hunk 14 above)
Comment (pzinemanas) [reply ingenialink/ethernet/dictionary.py]: Done!

File: ingenialink/virtual/virtual_drive.py (hunk 15 above)
Comment (pzinemanas) [reply ingenialink/virtual/virtual_drive.py]: Done!

File: mypy.ini (hunk 16 above)
Comment (pzinemanas) [reply mypy.ini:19]: Fixed!

File: ingenialink/dictionary.py (hunk 7 above)
Comment (GullinBustin) [reply ingenialink/dictionary.py:131]: Ok, I see, it is out of the issue scope, it's ok.

**Review by GullinBustin — APPROVED** [review]
